src/gui/dialogs/migrate_version_selection.cpp

```diff
@@ -97,19 +98,52 @@ void migrate_version_selection::post_show(window& window)
 		// given self-compilation and linux distros being able to do whatever they want plus command line options to
 		// alter locations make sure the directories/files are actually different before doing anything with them
 		if(migrate_addons_dir != filesystem::get_addons_dir()) {
+			std::vector<std::string> old_addons;
+			std::vector<std::string> current_addons;
 			std::vector<std::string> migrate_addons;
-			filesystem::get_files_in_dir(migrate_addons_dir, nullptr, &migrate_addons);
+
+			filesystem::get_files_in_dir(migrate_addons_dir, nullptr, &old_addons);
+			filesystem::get_files_in_dir(filesystem::get_addons_dir(), nullptr, &current_addons);
+
+			std::set_difference(old_addons.begin(), old_addons.end(), current_addons.begin(), current_addons.end(), std::back_inserter(migrate_addons));
+
 			if(migrate_addons.size() > 0) {
 				ad_hoc_addon_fetch_session(migrate_addons);
 			}
 		}
 
 		if(migrate_prefs_file != filesystem::get_prefs_file() && filesystem::file_exists(migrate_prefs_file)) {
-			filesystem::copy_file(migrate_prefs_file, filesystem::get_prefs_file());
+			// if this is the first time, just copy the file over
+			// else need to merge the preferences file
+			if(!filesystem::file_exists(filesystem::get_prefs_file())) {
+				filesystem::copy_file(migrate_prefs_file, filesystem::get_prefs_file());
+			} else {
+				config current_cfg;
+				read(current_cfg, filesystem::get_prefs_file());
+				config old_cfg;
+				read(old_cfg, migrate_prefs_file);
+
+				// when both files have the same attribute, use the one from whichever was most recently modified
+				bool current_prefs_are_older = filesystem::file_modified_time(filesystem::get_prefs_file()) < filesystem::file_modified_time(migrate_prefs_file);
+				for(const config::attribute& val : old_cfg.attribute_range()) {
+					if((current_cfg.has_attribute(val.first) && !current_prefs_are_older) || !current_cfg.has_attribute(val.first)) {
+						preferences::set(val.first, val.second);
```

Comment:
What happens when reloading the preferences afterwards? It seems like that would undo this unless the preferences are first saved or is that done implicitly somewhere?

Reply:
I don't know - I haven't tested this yet.

Reply:
Looks like it does use the values moved during the migration, which I'm actually surprised about.

Reply:
That does sound odd. If reloading does not overwrite the current settings then the case of copying the file does not work right? I don't see how both can work correctly.

Reply:
What does the parser's read() method do when passed a config that has existing data in it?

Reply:
The config is cleared.

https://github.com/wesnoth/wesnoth/blob/08ee0dddd1376ceeaa10431910c8d02f6393fb4d/src/serialization/parser.cpp#L632-L635

https://github.com/wesnoth/wesnoth/blob/08ee0dddd1376ceeaa10431910c8d02f6393fb4d/src/serialization/parser.cpp#L122-L124

Reply:
How does it work on other OS-es?

Reply:
Other platforms have the preferences as being per-version - otherwise there wouldn't be any comments from players about losing their campaign completions when updating.

Reply:
Alright, tested with:
* My clone of `wesnoth/wesnoth` set to 1.18 and using a config path that contained "1.18" in it.
* My clone of `wesnoth/Pentarctagon` set to 1.19 and using a config path that contained "1.19" in it.
* I tested having `UI_sound=yes` in 1.18 and `UI_sound=no` in 1.19, with the 1.18 preferences being newer. After doing the migration, this correctly resulted in 1.19's preferences being set to have `UI_sound=yes`.
* I tested having `UI_sound=yes` in 1.18 and `UI_sound=no` in 1.19, with the 1.19 preferences being newer. After doing the migration, this correctly resulted in 1.19's preferences still having `UI_sound=no` post-migration.

The call to `preferences::write_preferences()` was also needed for this to work properly. Also important for me to remember for next time: the parser's read() overload that takes a string isn't the path to a file, it's the string contents of a config.